awx/main/scheduler/dag_workflow.py

```diff
@@ -63,40 +112,113 @@ def cancel_node_jobs(self):
         return cancel_finished
 
     def is_workflow_done(self):
-        root_nodes = self.get_root_nodes()
-        nodes = root_nodes
-        is_failed = False
+        for node in self.nodes:
+            obj = node['node_object']
+            if obj.do_not_run is False and not obj.job and obj.unified_job_template:
+                return False
+            elif obj.job and obj.job.status not in ['successful', 'failed', 'canceled', 'error']:
+                return False
+        return True
 
-        for index, n in enumerate(nodes):
-            obj = n['node_object']
-            job = obj.job
+    def has_workflow_failed(self):
+        failed_nodes = []
+        res = False
+        failed_path_nodes_id_status = []
+        failed_unified_job_template_node_ids = []
 
-            if obj.unified_job_template is None:
-                is_failed = True
-                continue
-            elif not job:
-                return False, False
-
-            children_success = self.get_dependencies(obj, 'success_nodes')
-            children_failed = self.get_dependencies(obj, 'failure_nodes')
-            children_always = self.get_dependencies(obj, 'always_nodes')
-            if not is_failed and job.status != 'successful':
-                children_all = children_success + children_failed + children_always
-                for child in children_all:
-                    if child['node_object'].job:
-                        break
+        for node in self.nodes:
+            obj = node['node_object']
+            if obj.do_not_run is False and obj.unified_job_template is None:
+                failed_nodes.append(node)
+            elif obj.job and obj.job.status in ['failed', 'canceled', 'error']:
+                failed_nodes.append(node)
+
+        for node in failed_nodes:
+            obj = node['node_object']
+            if (len(self.get_dependencies(obj, 'failure_nodes')) +
+                    len(self.get_dependencies(obj, 'always_nodes'))) == 0:
```

Comment:
it's a little strange that an always node path is counted as a valid error handler. This should be in docs/changelog.